[Bug]: The Safari instance is already paired with another WebDriver session

```java
Driver Factory Action "createNewLocalDriverInstance" failed. With the following test data "Failed to create new Browser Session". Root cause: "org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 500. Message: Could not create a session: The Safari instance is already paired with another WebDriver session. "

      java.lang.AssertionError: Driver Factory Action "createNewLocalDriverInstance" failed. With the following test data "Failed to create new Browser Session". Root cause: "org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 500. Message: Could not create a session: The Safari instance is already paired with another WebDriver session. "
at org.testng.Assert.fail(Assert.java:99)
at com.shaft.tools.io.internal.FailureReporter.fail(FailureReporter.java:38)
at com.shaft.driver.internal.DriverFactory.DriverFactoryHelper.failAction(DriverFactoryHelper.java:110)
at com.shaft.driver.internal.DriverFactory.DriverFactoryHelper.createNewLocalDriverInstance(DriverFactoryHelper.java:379)
at com.shaft.driver.internal.DriverFactory.DriverFactoryHelper.initializeDriver(DriverFactoryHelper.java:598)
at com.shaft.driver.internal.DriverFactory.DriverFactoryHelper.initializeDriver(DriverFactoryHelper.java:554)
at com.shaft.driver.DriverFactory.getHelper(DriverFactory.java:109)
at com.shaft.driver.SHAFT$GUI$WebDriver.<init>(SHAFT.java:42)
at testPackage.ChainableElementActionsTests.beforeMethod(ChainableElementActionsTests.java:15)
at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
at java.base/java.lang.reflect.Method.invoke(Method.java:580)
at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
at org.testng.internal.invokers.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:69)
at org.testng.internal.invokers.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:393)
at org.testng.internal.invokers.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:326)
at org.testng.internal.invokers.TestInvoker.runConfigMethods(TestInvoker.java:810)
at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:578)
at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:228)
at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:63)
at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:961)
at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:201)
at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:148)
at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
at org.testng.internal.thread.graph.TestNGFutureTask.run(TestNGFutureTask.java:22)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 500. Message: Could not create a session: The Safari instance is already paired with another WebDriver session. 
Host info: host: 'Mac-1711185861525.local', ip: 'fe80:0:0:0:ca9:270c:8120:39f1%en0'
Build info: version: '4.18.1', revision: 'b1d3319b48'
System info: os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '12.7.3', java.version: '21.0.2'
Driver info: org.openqa.selenium.safari.SafariDriver
Command: [null, newSession {capabilities=[Capabilities {browserName: safari, pageLoadStrategy: normal, timeouts: {pageLoad: 60000, script: 60000}, unhandledPromptBehavior: ignore}]}]
Capabilities {browserName: safari, pageLoadStrategy: normal, timeouts: {pageLoad: 60000, script: 60000}, unhandledPromptBehavior: ignore}
at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:139)
at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:95)
at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:67)
at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:162)
at org.openqa.selenium.remote.service.DriverCommandExecutor.invokeExecute(DriverCommandExecutor.java:216)
at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:174)
at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:519)
at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:233)
at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:162)
at org.openqa.selenium.safari.SafariDriver.<init>(SafariDriver.java:84)
at org.openqa.selenium.safari.SafariDriver.<init>(SafariDriver.java:80)
at org.openqa.selenium.safari.SafariDriver.<init>(SafariDriver.java:57)
at com.shaft.driver.internal.DriverFactory.DriverFactoryHelper.createNewLocalDriverInstance(DriverFactoryHelper.java:305)
... 25 more
```

--- a/shaft-engine/src/main/java/com/shaft/driver/internal/DriverFactory/DriverFactoryHelper.java
+++ b/shaft-engine/src/main/java/com/shaft/driver/internal/DriverFactory/DriverFactoryHelper.java
@@ -1007,14 +1007,14 @@ public class DriverFactoryHelper {
                     }
                 } else if (message != null && message.contains("The Safari instance is already paired with another WebDriver session.")) {
                     //this issue happens when running locally via safari/mac platform
-                    // attempting blind fix by trying to quit existing safari instances if any
+                    // attempting blind fix by trying to quit existing safari instances if any, then
+                    // retrying (issue #1548): safaridriver only allows one active session system-wide,
+                    // so killing the stale pairing is useless unless we retry with the now-freed instance
                     try {
                         SHAFT.CLI.terminal().performTerminalCommands(Arrays.asList(
                                 "osascript -e 'quit app \"Safari\"'", "osascript -e 'quit app \"SafariDriver\"'",
                                 "pkill -x Safari", "pkill -x SafariDriver",
                                 "killall Safari", "killall SafariDriver"));
-                        //minimizing retry attempts to save execution time
-                        shouldRetry = false;
                     } catch (Throwable throwable) {
                         ReportManagerHelper.logDiscrete(throwable, Level.DEBUG);
                     }

--- a/shaft-engine/src/test/java/com/shaft/driver/internal/DriverFactory/DriverFactoryHelperCoverageUnitTest.java
+++ b/shaft-engine/src/test/java/com/shaft/driver/internal/DriverFactory/DriverFactoryHelperCoverageUnitTest.java
@@ -1,5 +1,6 @@
 package com.shaft.driver.internal.DriverFactory;
 
+import com.shaft.cli.TerminalActions;
 import com.shaft.driver.SHAFT;
 import com.shaft.driver.internal.DriverFactory.DriverFactoryHelper;
 import com.shaft.driver.internal.DriverFactory.OptionsManager;
@@ -34,6 +35,7 @@ import org.openqa.selenium.remote.RemoteWebDriver;
 import org.openqa.selenium.remote.http.ClientConfig;
 import org.openqa.selenium.safari.SafariOptions;
 import org.openqa.selenium.safari.SafariDriver;
+import org.openqa.selenium.SessionNotCreatedException;
 import io.github.bonigarcia.wdm.WebDriverManager;
 import org.mockito.MockedConstruction;
 import org.testng.annotations.AfterMethod;
@@ -1203,6 +1205,39 @@ public class DriverFactoryHelperCoverageUnitTest {
             helper.initializeDriver(com.shaft.driver.DriverFactory.DriverType.CHROME, null);
             org.mockito.Mockito.verify(window).setSize(new org.openqa.selenium.Dimension(1234, 678));
             SHAFT.Validations.assertThat().object(helper.getDriver()).isEqualTo(mockedRemoteDriver).perform();
+        }
+    }
+
+    @Test
+    public void createNewLocalDriverInstanceShouldRetryLocalSafariSessionAfterRecoveringFromPairingConflict() throws Exception {
+        // issue #1548: safaridriver permits exactly one active WebDriver session system-wide. When
+        // that "already paired" conflict is hit, the recovery branch kills any lingering
+        // Safari/SafariDriver processes -- but it must then retry driver construction, or the
+        // recovery is pointless: the session gets freed but the caller still fails with the very
+        // error the recovery just fixed.
+        DriverFactoryHelper helper = new DriverFactoryHelper();
+        OptionsManager mockedOptionsManager = org.mockito.Mockito.mock(OptionsManager.class);
+        org.mockito.Mockito.when(mockedOptionsManager.getSfOptions())
+                .thenThrow(new SessionNotCreatedException(
+                        "Could not create a session: The Safari instance is already paired with another WebDriver session."))
+                .thenReturn(new SafariOptions());
+
+        Field optionsManagerField = DriverFactoryHelper.class.getDeclaredField("optionsManager");
+        optionsManagerField.setAccessible(true);
+        optionsManagerField.set(helper, mockedOptionsManager);
+
+        Method createNewLocalDriverInstance = DriverFactoryHelper.class.getDeclaredMethod(
+                "createNewLocalDriverInstance", com.shaft.driver.DriverFactory.DriverType.class, int.class);
+        createNewLocalDriverInstance.setAccessible(true);
+
+        try (MockedConstruction<TerminalActions> ignoredTerminal = org.mockito.Mockito.mockConstruction(TerminalActions.class);
+             MockedConstruction<SafariDriver> ignoredSafari = org.mockito.Mockito.mockConstruction(SafariDriver.class)) {
+            createNewLocalDriverInstance.invoke(helper, com.shaft.driver.DriverFactory.DriverType.SAFARI, 1);
+
+            SHAFT.Validations.assertThat().object(helper.getDriver()).isNotNull().perform();
+            org.mockito.Mockito.verify(mockedOptionsManager, org.mockito.Mockito.times(2)).getSfOptions();
+            org.mockito.Mockito.verify(ignoredTerminal.constructed().get(0))
+                    .performTerminalCommands(org.mockito.ArgumentMatchers.anyList());
         }
     }
 

